lib/vnet/ssh_handler.go

```diff
@@ -92,37 +96,121 @@ func (h *sshHandler) handleTCPConnectorWithTargetConn(
+		initiatedSSHConn bool
+	)
+	sendAuthBanner := func(msg string) {
+		// This is super hacky but:
+		// - Auth banners need to be sent before the authentication phase
+		//   completes.
+		// - As soon as we return from PublicKeyCallback, x/crypto/ssh will
+		//   complete the authentication phase.
+		// - Therefore we need to call SendAuthBanner before returning
+		//   from PublicKeyCallback.
+		// - However, SendAuthBanner will block until the authentication phase
+		//   completes so we must call it from a goroutine.
+		// - There is a race: x/crypto/ssh may complete the authentication phase
+		//   before SendAuthBanner gets a chance to enqueue the banner message.
+		// - So it seems the best we can do is sleep before returning with the
+		//   hope that SendAuthBanner gets a chance to run before x/crypto/ssh
+		//   completes the authentication phase.
+		// - There is still a logical race, but the auth banner is not critical
+		//   so it seems okay to consider this a best-effort attempt.
+		go preAuthConn.SendAuthBanner("VNet: " + msg + "\n")
+		time.Sleep(100 * time.Millisecond)
+	}
 	serverConfig := &ssh.ServerConfig{
+		PreAuthConnCallback: func(conn ssh.ServerPreAuthConn) {
+			preAuthConn = conn
+		},
+		// We attempt to initiate an SSH connection with the target server in
+		// PublicKeyCallback in order to fail the SSH authentication phase with
+		// the client if SSH authentication to the target fails. Otherwise, when
+		// connection to an SSH node the user is not allowed to access, they
+		// would just see an succesfull SSH handshake and then an immediately
+		// closed connection.
+		//
+		// TODO(nklaassen): if https://github.com/golang/go/issues/70795 ever
+		// gets implemented we should do this in VerifiedPublicKeyCallback
+		// instead.
 		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
 			if !sshutils.KeysEqual(h.cfg.sshProvider.trustedUserPublicKey, key) {
-				return nil, trace.AccessDenied("SSH client public key is not trusted")
+				return nil, trace.AccessDenied("client public key is not trusted")
+			}
+			// Make sure to only initiate the SSH connection once in case
+			// PublicKeyCallback is called multiple times.
+			if initiatedSSHConn {
+				return &ssh.Permissions{}, clientConnErr
+			}
+			initiatedSSHConn = true
+			clientConn, clientConnErr = h.initiateSSHConn(ctx, targetConn, conn.User())
+			if clientConnErr != nil {
+				// Attempt to send a friendlier errer message if we failed to
+				// initiate the SSH connection to the target by sending an auth
+				// banner message.
+				if utils.IsHandshakeFailedError(clientConnErr) {
+					// We don't have much real information about the error in
+					// this case, this is the same message tsh prints.
+					sendAuthBanner(fmt.Sprintf("access denied to %s connecting to %s", conn.User(), target.hostname))
+				} else {
+					sendAuthBanner(trace.UserMessage(clientConnErr))
+				}
+				return nil, clientConnErr
```

Comment:
We could use a `ssh.BannerError` rather than using `ssh.ServerPreAuthConn` directly, it's a bit simpler.